pkg/storage/openstack.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// ReauthRoundTripper satisfies the http.RoundTripper interface and is used to
+// limit the number of re-auth attempts (infinite by default)
+type ReauthRoundTripper struct {
+	rt                http.RoundTripper
+	numReauthAttempts int
+}
+
+// RoundTrip performs a round-trip HTTP request and logs relevant information about it.
+func (rrt *ReauthRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	response, err := rrt.rt.RoundTrip(request)
+	if response == nil {
+		return nil, err
+	}
+
+	if response.StatusCode == http.StatusUnauthorized {
+		if rrt.numReauthAttempts == 3 {
+			return response, errors.New("tried to re-authenticate 3 times with no success")
+		}
+		rrt.numReauthAttempts++
+	}
+
+	return response, nil
+}
+
+// OpenstackOSBackend is a storage backend for Openstack Object Storage
+type OpenstackOSBackend struct {
+	Container string
+	Prefix    string
+	Region    string
+	CACert    string
+	Client    *gophercloud.ServiceClient
+}
+
+// NewOpenstackOSBackend creates a new instance of OpenstackOSBackend
+func NewOpenstackOSBackend(container string, prefix string, region string, caCert string) *OpenstackOSBackend {
+	authOptions, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		panic(fmt.Sprintf("Openstack (environment): %s", err))
+	}
+	authOptions.AllowReauth = true
+
+	// Create a custom HTTP client to handle reauth retry and custom CACERT if needed
+	roundTripper := ReauthRoundTripper{}
+	if caCert != "" {
+		caCert, err := ioutil.ReadFile(caCert)
+		if err != nil {
+			panic(fmt.Sprintf("Openstack (ca certificates): %s", err))
+		}
+
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+
+		transport := &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: caCertPool,
+			},
+		}
+		roundTripper.rt = transport
+	} else {
+		roundTripper.rt = http.DefaultTransport
+	}
+
+	provider, err := openstack.NewClient(authOptions.IdentityEndpoint)
+	if err != nil {
+		panic(fmt.Sprintf("Openstack (client): %s", err))
+	}
+
+	provider.HTTPClient = http.Client{
+		Transport: &roundTripper,
+	}
+
+	err = openstack.Authenticate(provider, authOptions)
+	if err != nil {
+		panic(fmt.Sprintf("Openstack (authenticate): %s", err))
+	}
+
+	client, err := openstack.NewObjectStorageV1(provider, gophercloud.EndpointOpts{
+		Region: region,
+	})
+	if err != nil {
+		panic(fmt.Sprintf("Openstack (object storage): %s", err))
+	}
+
+	b := &OpenstackOSBackend{
+		Container: container,
+		Prefix:    prefix,
+		Region:    region,
+		Client:    client,
+	}
+
+	return b
+}
+
+// ListObjects lists all objects in an Openstack container, at prefix
+func (b OpenstackOSBackend) ListObjects(prefix string) ([]Object, error) {
+	var objects []Object
+
+	prefix = pathutil.Join(b.Prefix, prefix)
+	opts := &osObjects.ListOpts{
+		Full:   true,
+		Prefix: prefix,
+	}
+
+	pager := osObjects.List(b.Client, b.Container, opts)
+	page, err := pager.AllPages()
```

Comment:
Can we introduce a kind of iteration here ? What happens if there are thousands of charts ?

Reply:
Can sure do something like that, it's basically what the client is doing in AllPages() but I imagine it would be more memory efficient to do it on my side. Will look into that.